back/engines/free/user_confirmation/app/models/user_confirmation/extensions/user.rb

```diff
@@ -0,0 +1,66 @@
+module UserConfirmation
+  module Extensions
+    module User
+      def self.included(base)
+        base.class_eval do
+          validates :email_confirmation_code, format: { with: USER_CONFIRMATION_CODE_PATTERN }, allow_nil: true
+          validates :email_confirmation_retry_count, numericality: { less_than_or_equal_to: ENV.fetch('EMAIL_CONFIRMATION_MAX_RETRIES', 5) }
+          validates :email_confirmation_code_reset_count, numericality: { less_than_or_equal_to: ENV.fetch('EMAIL_CONFIRMATION_MAX_RETRIES', 5) }
+
+          before_validation :reset_confirmation_code, unless: :email_confirmation_code, on: :create
+        end
+      end
+
+      def requires_confirmation?
+        !(active? || invited? || confirmed? || registered_with_phone?)
+      end
+
+      def confirmed?
+        email_confirmed_at.present?
+      end
+
+      def confirm!
+        return false unless registered_with_email?
+
+        self.email_confirmed_at = Time.zone.now
+        save
```

Comment:
I would also suggest `save!` here, especially because the `confirm!` method also ends with a `!`, so that's what I would expect from its behaviour.

Reply:
Its a good point, by adding `!` I was addressing the ruby pattern of destructive vs non-destructive patterns, e.g. `select` vs `select!`, in which the lattern modifies the original object.